test/integration/targets/mongodb_shard/tasks/main.yml

```diff
@@ -0,0 +1,367 @@
+     - "localhost:3006"
+
+- name: Ensure /home/tests/config dir exists
+  file:
+    path: /home/tests/config
+    state: directory
+
+- name: Launch cfg server
+  command: mongod --configsvr --port 4000 --dbpath /home/tests/config --logpath /home/tests/config.log --smallfiles --replSet "{{ configsrv_replicaset }}" --fork
+
+- name: Create config srv replicaset with module
+  mongodb_replicaset:
+    login_user: "{{ mongodb_admin_user }}"
+    login_password: "{{ mongodb_admin_password }}"
+    login_port: 4000
+    login_database: "admin"
+    replica_set: "{{ configsrv_replicaset }}"
+    validate: no
+    members:
+     - "localhost:4000"
+
+- pause:
+    seconds: 30
+
+- name: Add mongodb admin user to each shard 3.4+
+  mongodb_user:
+    login_host: localhost
+    login_port: "{{ item.port }}"
+    replica_set: "{{ item.rs }}"
+    database: admin
+    name: "{{ mongodb_admin_user }}"
+    password: "{{ mongodb_admin_password }}"
+    roles: ["root"]
+    state: present
+  register: mongo_admin_user
+  with_items:
+    - { "port": 3001, "rs": "{{ mongodb_replicaset1 }}" }
+    - { "port": 3004, "rs": "{{ mongodb_replicaset2 }}" }
+    - { "port": 4000, "rs": "{{ configsrv_replicaset }}" }
+  when: mongodb_version not in ["3.2", "4.0"] and test_mongo_auth == True
+
+- name: Add mongodb admin user to each shard 3.2 ^ 4.0
+  mongodb_user:
+    login_host: localhost
+    login_port: "{{ item.port }}"
+    replica_set: "{{ item.rs }}"
+    database: admin
+    name: "{{ mongodb_admin_user }}"
+    password: "{{ mongodb_admin_password }}"
+    roles: ["root"]
+    state: present
+  register: mongo_admin_user
+  with_items:
+    - { "port": 3001, "rs": "{{ mongodb_replicaset1 }}" }
+    - { "port": 3004, "rs": "{{ mongodb_replicaset2 }}" }
+  when: mongodb_version not in ["3.2", "4.0"] and test_mongo_auth == True
+
+# mongodb_user throws an error when creating a user on 3.2 (also on 4.0 with Ubuntu 18.04)
+# 'majority' is the only valid write concern when writing to config server replica sets
+- name: Copy create_user_root_3.2.js.j2 template to host
+  template:
+    src: create_user_root_3.2.js.j2
+    dest: /root/create_user_root_3.2.js
+  when: mongodb_version in ["3.2", "4.0"]
+
+- name: Create admin user on 3.2 and 4.0 config replset
+  shell: mongo admin --port 4000 /root/create_user_root_3.2.js
+  when: mongodb_version in ["3.2", "4.0"]
+
+- name: Murder all mongod processes
+  shell: pkill mongod && sleep 10 || sleep 10;
+
+- name: Execute mongod script to launch with auth shard1
+  script: files/bash/mongodb_replicaset.sh 3000 "{{ mongodb_replicaset1 }}" 3 1
+
+- name: Execute mongod script to launch with auth shard2
+  script: files/bash/mongodb_replicaset.sh 3003 "{{ mongodb_replicaset2 }}" 3 1
+
+- name: Launch cfg server with auth
+  command: mongod --configsvr --port 4000 --dbpath /home/tests/config --logpath /home/tests/config.log --smallfiles --replSet "{{ configsrv_replicaset }}" --fork --auth --keyFile /root/my.key
+
+- pause:
```

Comment:
Integration tests should not pause. This integration test has over 4 minutes of pauses in it.

Reply:
@Mattclay They need to pause to allow process to become active and elections to happen. The alternative would be to continually poll the replicaset status  and we effectively still pause anyway.

Reply:
Polling would be ideal, perhaps with `wait_for` if possible. That way the tests can continue as soon as the services are ready instead of waiting the full 10 seconds. Getting the test run time down is important, and polling will hopefully make a significant difference.

Reply:
Polling has been replaced in the majority of instances. I use a js script to poll the mongod instance and return once it's in the primary status. Seem to work well.

Reply:
A few pauses remain. Mainly to wait for processes to fully die after tearing down the mongodb replicasets.